pkg/oc/admin/diagnostics/diagnostics.go

```diff
@@ -82,50 +80,100 @@ const (
 
 		    %[1]s --master-config=/etc/origin/master/master-config.yaml
 
-		* If master/node config files are not found and the --host flag is not
-		  present, host diagnostics are skipped.
-		* If the client has cluster-admin access, this access enables cluster
-		  diagnostics to run which regular users cannot.
-		* If a client config file is not found, client and cluster diagnostics
-		  are skipped.
+		* Explicitly specifying a config file raises an error if it is not found.
+		* A client config with cluster-admin access is required for most cluster diagnostics.
+		* Diagnostics that require a config file are skipped if it is not found.
+		* The standard set also skips diagnostics considered too heavyweight.
 
-		Diagnostics may be individually run by passing diagnostic name as arguments.
+		An individual diagnostic may be run as a subcommand which may have flags
+		for specifying options specific to that diagnostic.
+
+		Finally, the "all" subcommand runs all available diagnostics (including heavyweight
+		ones skipped in the standard set) and provides all individual diagnostic flags.
+		`)
+	longDescriptionAll = templates.LongDesc(`
+		This utility helps troubleshoot and diagnose known problems for an OpenShift cluster
+		and/or local host. This subcommand exists to run all available diagnostics:
+
+		    %[1]s
 
-		    %[1]s <DiagnosticName>
+		Available diagnostics vary based on client config and local OpenShift host config.
+		All flags from the base command work similarly here, but all possible flags for
+		individual diagnostics are also available.
+		`)
+	longDescriptionIndividual = templates.LongDesc(`
+		Runs the %s diagnostic.
 
-		The available diagnostic names are: %[2]s.`)
+		%s
+		`)
 )
 
 // NewCmdDiagnostics is the base command for running any diagnostics.
 func NewCmdDiagnostics(name string, fullName string, out io.Writer) *cobra.Command {
-	o := &DiagnosticsOptions{
-		RequestedDiagnostics: []string{},
-		LogOptions:           &log.LoggerOptions{Out: out},
-		ImageTemplate:        variable.NewDefaultImageTemplate(),
-		NetworkOptions:       &NetworkDiagnosticsOptions{},
+	available := availableDiagnostics()
+	o := &DiagnosticsConfig{
+		RequestedDiagnostics:     available.Names().Difference(defaultSkipDiagnostics()),
+		ParameterizedDiagnostics: types.NewParameterizedDiagnosticMap(available...),
+		LogOptions:               &log.LoggerOptions{Out: out},
 	}
 
 	cmd := &cobra.Command{
 		Use:   name,
 		Short: "Diagnose common cluster problems",
-		Long:  fmt.Sprintf(longDescription, fullName, strings.Join(availableDiagnostics().List(), ", ")),
+		Long:  fmt.Sprintf(longDescription, fullName),
 		Run: func(c *cobra.Command, args []string) {
-			kcmdutil.CheckErr(o.Complete(args))
+			kcmdutil.CheckErr(o.Complete(c, args))
+
+			failed, err, warnCount, errorCount := o.RunDiagnostics()
+			o.Logger.Summary(warnCount, errorCount)
+
+			kcmdutil.CheckErr(err)
+			if failed {
```

Comment:
Is there any case where `failed` will be true but you get a nil `err`? Otherwise you don't need this check.

Reply:
There is actually; this is meant to distinguish between problems encountered while gathering everything the diagnostics need (errors as we usually think of them), versus the the problems they find and report (failures which still need to be reflected in the return code).

However, I think it may actually be possible and clearer if those two stages were extracted. If it's not too messy, I'll try including it in this change.

Reply:
So I guess it's no surprise - it's messy. And I'd like to improve that; there just has to be a way for all this code to be more legible.

To revise my earlier statement, there is more than one error mode here. There are errors in constructing diagnostics that need to be reported to the user without halting execution. There are conditions under which we should abort without running diagnostics at all. And there are diagnostic results that indicate problems to report with an error exit code. I need to think more about the best way to signal which mode is of concern in each place; `ok` and `err` and `failure` just don't provide much nuance.

But probably not for this patch set.